### ISSUE TITLE
[macOS] Fails to launch: Task Manager

### ISSUE THREAD

**Issue by shinra-electric** [Swartz27/xray-16#1731]
**Describe the bug**
Game fails to launch on macOS when the source code folder is not present.

**To Reproduce**
I usually create an app bundle for OpenXRay using [this script](https://github.com/shinra-electric/OpenXRay-Build-Script).
After the app is bundled, the script will delete the xray-16 source code folder.

When trying to launch the game, it will fail to launch because it cannot find `xray-16/src/xrCore/xrDebug.cpp`, which no longer exists. 

Additional information:
`xrCore.dylib` is present in the app bundle, and the `xr_3da` executable is able to find it there. 

Terminal log:
[Terminal Saved Output.txt](https://github.com/user-attachments/files/17700354/Terminal.Saved.Output.txt)

**Desktop:**
 - OS: macOS 15.1 Sequoia
 - Building directly from dev branch (11th November 2024)

**Comment by Xottab-DUTY** [comment]
No, xrDebug.cpp is not the crash reason. I'm quite not sure what the real reason is.

Do you have the game files in `/Users/user/Library/Application Support/GSC Game World/S.T.A.L.K.E.R. - Call of Pripyat/` folder?
It's a hardcoded path, where the game should be.

**Comment by shinra-electric** [comment]
> Do you have the game files in `/Users/user/Library/Application Support/GSC Game World/S.T.A.L.K.E.R. - Call of Pripyat/` folder? It's a hardcoded path, where the game should be.

Yes, the script copies over the data folders if it detects them.

**Comment by Xottab-DUTY** [comment]
> objc[864]: Class SDLApplication is implemented in both /opt/homebrew/Cellar/sdl2/2.30.9/lib/libSDL2-2.0.0.dylib (0x1039a46c8) and /Applications/Games/OpenXRay.app/Contents/libs/libSDL2-2.0.0.dylib (0x10332c6c8). One of the two will be used. Which one is undefined.

Hmm, if it's not the problem with files, then this might also be related. What if you delete `/Applications/Games/OpenXRay.app/Contents/libs/libSDL2-2.0.0.dylib` so that the engine would use system's dylib?

**Comment by shinra-electric** [comment]
> Hmm, if it's not the problem with files, then this might also be related. What if you delete `/Applications/Games/OpenXRay.app/Contents/libs/libSDL2-2.0.0.dylib` so that the engine would use system's dylib?

I know the reason for this warning, but it's usually benign. 

I built it again, but just normally without making a bundle and keeping the source folder. The error still occurs.

Here is the terminal output:
~~[OpenXRay_Output_CLI.txt]()~~
Sorry, use this one instead: 
[OpenXRay_Output_CLI2.txt](https://github.com/user-attachments/files/17700993/OpenXRay_Output_CLI2.txt)


I included the CMake config and build output, just in case there is anything relevant there.

**Comment by Xottab-DUTY** [comment]
At first I didn't paid attention to this:

> stack trace:
> 
> xrDebug::GatherInfo(char*, unsigned long, ErrorLocation const&, char const*, char const*, char const*, char const*)
> xr_terminate()
> std::__terminate(void (*)())
> std::terminate()
> std::__1::condition_variable::__do_timed_wait(std::__1::unique_lock<std::__1::mutex>&, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)
> TaskManager::TaskWorkerStart()
> void* std::__1::__thread_proxy[abi:ne180100]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, std::__1::thread Threading::RunThread<void (TaskManager::*)(), TaskManager*>(char const*, void (TaskManager::*&&)(), TaskManager*&&)::'lambda'(void (TaskManager::*&&)(), TaskManager*&&), void (TaskManager::*)(), TaskManager*>>(void*)
> 7   libsystem_pthread.dylib             0x0000000192af32e4 _pthread_start + 136
> 8   libsystem_pthread.dylib             0x0000000192aee0fc thread_start + 8

Seems like some problem with our task manager.
Will try to investigate and I hope that I it's possible for me to fix it (I don't have any Mac 🙁)

**Comment by Xottab-DUTY** [comment]
@shinra-electric, can you run it under the debugger (e.g. GDB)?

**Comment by shinra-electric** [comment]
> @shinra-electric, can you run it under the debugger (e.g. GDB)?

GDB doesn't run on Arm. What would be a good alternative?

**Comment by Xottab-DUTY** [comment]
> > @shinra-electric, can you run it under the debugger (e.g. GDB)?
> 
> GDB doesn't run on Arm. What would be a good alternative?

Maybe LLDB?

**Comment by shinra-electric** [comment]
I opened the executable in Xcode, which uses lldb. Here is the log: 

[openxray_lldb_output.txt](https://github.com/user-attachments/files/17731418/openxray_lldb_output.txt)
![Screenshot 2024-11-13 at 11 10 31](https://github.com/user-attachments/assets/2af5263e-c17f-42c1-b0e5-fcc6c36e6c84)

**Comment by Xottab-DUTY** [comment]
It looks that it isn't easy to solve. I'm not sure how.
After looking on the internet, it seems like a macOS bug with C++17 support, not ours.

Help is wanted and appreciated!

**Comment by Xottab-DUTY** [comment]
> After looking on the internet, it seems like a macOS bug with C++17 support, not ours.

It was our bug actually. Please, open a new issue if any other crash with task manager appears!

**Comment by shinra-electric** [comment]
> It was our bug actually. Please, open a new issue if any other crash with task manager appears!

Nice, this seems to have fixed the system freezes. I haven't had any freezes after running the game several times. 

There is still a crash on quit, but it's not a big deal and probably unrelated.

**Comment by Xottab-DUTY** [comment]
> There is still a crash on quit, but it's not a big deal and probably unrelated.

I just (literally 2 minutes ago) fixed #1772, it should be better now.

**Comment by shinra-electric** [comment]
> I just (literally 2 minutes ago) fixed [#1772](https://github.com/OpenXRay/xray-16/issues/1772), it should be better now.

There is still a segfault when quitting from in-game using Command-Q on macOS. 
Quitting with Command-Q on the main menu before going in-game works fine.
Quitting from the menu using the `Exit to Windows` button works fine at any time.  

[OXR Terminal log.txt](https://github.com/user-attachments/files/18468803/OXR.Terminal.log.txt)

But this is minor. I'm pretty happy the game is actually playable on macOS now.
